modules/flist/flist.go

```diff
@@ -0,0 +1,328 @@
+		return fmt.Errorf("specified path is not a directory")
+	}
+
+	if !strings.HasPrefix(path, f.root) {
+		return fmt.Errorf("trying to unmount a directory outside of the flist module boundaries")
+	}
+
+	if err := syscall.Unmount(path, syscall.MNT_DETACH); err != nil {
+		log.Error().Err(err).Str("path", path).Msg("fail to umount flist")
+	}
+	_, name := filepath.Split(path)
+	pidPath := filepath.Join(f.pid, name) + ".pid"
+	if err := waitPidFile(time.Second*2, pidPath, false); err != nil {
+		log.Error().Err(err).Str("path", path).Msg("0-fs daemon did not stopped properly")
+		return err
+	}
+
+	// clean up working dirs
+	logPath := filepath.Join(f.log, name) + ".log"
+	backend := filepath.Join(f.root, "backend", name)
+	for _, path := range []string{logPath, backend, path} {
+		if err := os.RemoveAll(path); err != nil {
+			log.Error().Err(err).Msg("fail to remove %s")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// downloadFlist downloads an flits from a URL
+// if the flist location also provide and md5 hash of the flist
+// this function will use it to avoid downloading an flist that is
+// already present locally
+func (f *flistModule) downloadFlist(url string) (string, error) {
+	// first check if the md5 of the flist is available
+	md5URL := url + ".md5"
+	resp, err := http.Get(md5URL)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		hash, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+
+		flistPath := filepath.Join(f.flist, strings.TrimSpace(string(hash)))
+		_, err = os.Stat(flistPath)
+		if err != nil && !os.IsNotExist(err) {
+			return "", err
+		}
+		if err == nil {
+			log.Info().Str("url", url).Msg("flist already in cache")
+			// flist is already present locally, just return its path
+			return flistPath, nil
+		}
+	}
+
+	log.Info().Str("url", url).Msg("flist not in cache, downloading")
+	// we don't have the flist locally yet, let's download it
+	resp, err = http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("fail to download flist: %v", resp.Status)
+	}
+
+	return f.saveFlist(resp.Body)
+}
+
+// saveFlist save the flist contained in r
+// it save the flist by its md5 hash
+// to avoid loading the full flist in memory to compute the hash
+// it uses a MultiWriter to write the flist in a temporary file and fill up
+// the md5 hash then it rename the file to the hash
+func (f *flistModule) saveFlist(r io.Reader) (string, error) {
+	tmp, err := ioutil.TempFile(f.flist, "*_flist_temp")
+	if err != nil {
+		return "", err
+	}
+	defer tmp.Close()
+
+	h := md5.New()
+	mr := io.MultiWriter(tmp, h)
+	if _, err := io.Copy(mr, r); err != nil {
+		return "", err
+	}
+
+	hash := fmt.Sprintf("%x", h.Sum(nil))
+	path := filepath.Join(f.flist, hash)
+	if err := os.MkdirAll(filepath.Dir(path), 0770); err != nil {
+		return "", err
+	}
+
+	if os.Rename(tmp.Name(), path); err != nil {
```

Comment:
os.Rename will not work across devices. since /tmp is usually a different partition or filesystem other than the final flist destination witch is usually a permanent disk partition (or cache) Rename will  fail.

Copying a file though might cause a race condition if u start to mount 2 locations with same flist. This is why in core0 I used to lock the flist file on disk first before i start the download to make sure only one "mount" is downloading the file.

Reply:
Like explained in the previous comment `tmp` file  is not located in `/tmp`. It is created in the same directory as its final location, so the Rename is an atomic operation. So we don't need locks and are safe from race conditions.